src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx

```diff
@@ -590,30 +599,32 @@ namespace xclcpuemhal2 {
         if( !fp )
         {
           if(mLogStream.is_open()) mLogStream << __func__ << " failed to create temporary dlopen file" << std::endl;
-
-          if(memTopology)
-          {
-            delete []memTopology;
-            memTopology = NULL;
-          }
           return -1;
         }
         fwrite(sharedlib,sharedliblength,1,fp);
         fflush(fp);
         fclose(fp);
       }
-      if(memTopology)
+      if (memTopology.get() && connectvitybuf.get())
       {
-        const mem_topology* m_mem = (reinterpret_cast<const ::mem_topology*>(memTopology));
-        if(m_mem)
+        const mem_topology* m_mem = (reinterpret_cast<const ::mem_topology*>(memTopology.get()));
+        const connectivity* m_conn = (reinterpret_cast<const ::connectivity*>(connectvitybuf.get()));
+        if (m_mem && m_conn)
```

Comment:
yes, we have sw emulation drivers in both edge and pcie code base.

Reply:
Would it make sense to have this in a common class, function, files, whatever to avoid any repetition?
Otherwise, twice the code means twice the complexity, twice the bugs, the risk of getting out of sync, etc.